File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque

def is_collect(s):
    dict = {']':'[', ')':'(', '}':'{' }
    queue = deque()
    for c in s:
        if c in ['[', '(', '{']:
            queue.append(c)
        else:
            if not queue:
                return False
            
            if queue[-1] == dict[c]:
                queue.pop()
            else:
                return False

    if queue:
        return False
    return True

def solution(s):
    answer = 0
    for i in range(len(s)):
        check = s[i:] + s[:i]
        answer += is_collect(check)
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <stack>
#include <map>

using namespace std;
map<char, char> bracket = {{'}','{'}, {')', '('}, {']', '['}};

int solution(string s) {
    int answer = 0;
    string ss = s + s;
    
    if (s.length()%2 != 0)
        return 0;
    
    for (int i=0; i<s.length(); i++){
        stack<char> stk;
        
        for (int j=i; j<i+s.length(); j++){
            if (ss[j]=='{' or ss[j]=='(' or ss[j]=='[' or stk.empty())
                stk.push(ss[j]);
            else{
                if (stk.top()=='}' or stk.top()==')' or stk.top()==']')
                    break;
                if (bracket[ss[j]] == stk.top())
                    stk.pop();
                else
                    stk.push(ss[j]);
            }
        }
        if(stk.empty()) answer += 1;
    }
    return answer;
}